cylc/uiserver/resolvers.py

```diff
@@ -321,6 +323,71 @@ async def play(cls, workflows, args, workflows_mgr, log):
         await workflows_mgr.scan()
         return response
 
+    @staticmethod
+    async def enqueue(stream, queue):
+        async for line in stream:
+            await queue.put(line.decode())
+
+    @classmethod
+    async def cat_log(cls, workflow: Tokens, log, info, task=None, file=None):
+        """Calls `cat log`.
+
+        Used for log subscriptions.
+        """
+        full_workflow = workflow
+        if file and task:
+            full_workflow = Tokens(f"{workflow.workflow_id}//{task}")
+        cmd = ['cylc', 'cat-log']
+        if file:
+            cmd += ['-f', file]
+        cmd += ['-m', 't']
+        cmd.append(full_workflow.id)
+        log.info(f'$ {" ".join(cmd)}')
+
+        # For info, below subprocess is safe (uses shell=false by default)
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+        buffer: List[str] = []
+        queue: asyncio.Queue = asyncio.Queue()
+        # Farm out reading from stdout stream to a background task
+        # This is to get around problem where stream is not EOF until
+        # subprocess ends
+        enqueue_task = asyncio.create_task(cls.enqueue(proc.stdout, queue))
+        op_id = info.root_value
+        try:
+            while (op_id in info.context['sub_statuses']
+                    and info.context['sub_statuses'][op_id] != 'stop'):
```

Comment:
I don't think it's possible for `op_id` not to be in `sub_statuses`?

```suggestion
            while info.context['sub_statuses'][op_id] != 'stop':
```

Or if it is:

```suggestion
            while info.context['sub_statuses'].get(op_id) != 'stop':
```